File: plugin-rust-agent/src/main/java/jetbrains/buildServer/rust/CargoRunnerBuildService.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.rust;

import jetbrains.buildServer.RunBuildException;
import jetbrains.buildServer.agent.ToolCannotBeFoundException;
import jetbrains.buildServer.agent.runner.BuildServiceAdapter;
import jetbrains.buildServer.agent.runner.ProcessListener;
import jetbrains.buildServer.agent.runner.ProgramCommandLine;
import jetbrains.buildServer.rust.cargo.*;
import jetbrains.buildServer.rust.logging.CargoLoggerFactory;
import jetbrains.buildServer.rust.logging.CargoLoggingListener;
import jetbrains.buildServer.util.StringUtil;
import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Cargo runner service.
 */
public class CargoRunnerBuildService extends BuildServiceAdapter {
    private final Map<String, ArgumentsProvider> myArgumentsProviders;

    public CargoRunnerBuildService() {
        myArgumentsProviders = new HashMap<String, ArgumentsProvider>();
        myArgumentsProviders.put(CargoConstants.COMMAND_BENCH, new BenchArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_BUILD, new BuildArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_CLEAN, new CleanArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_DOC, new DocArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_LOGIN, new LoginArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_PACKAGE, new PackageArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_PUBLISH, new PublishArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_RUN, new RunArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_RUSTC, new RustcArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_RUSTDOC, new RustDocArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_TEST, new TestArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_UPDATE, new UpdateArgumentsProvider());
        myArgumentsProviders.put(CargoConstants.COMMAND_YANK, new YankArgumentsProvider());
    }

    @NotNull
    @Override
    public ProgramCommandLine makeProgramCommandLine() throws RunBuildException {
        final Map<String, String> parameters = getRunnerParameters();

        final String commandName = parameters.get(CargoConstants.PARAM_COMMAND);
        if (StringUtil.isEmpty(commandName)) {
            RunBuildException buildException = new RunBuildException("Cargo command name is empty");
            buildException.setLogStacktrace(false);
            throw buildException;
        }

        final ArgumentsProvider argumentsProvider = myArgumentsProviders.get(commandName);
        if (argumentsProvider == null) {
            RunBuildException buildException = new RunBuildException("Unable to construct arguments for cargo command " + commandName);
            buildException.setLogStacktrace(false);
            throw buildException;
        }

        final String toolPath;
        final List<String> arguments = argumentsProvider.getArguments(parameters);
        try {
            toolPath = getToolPath(CargoConstants.RUNNER_TYPE);
        } catch (ToolCannotBeFoundException e) {
            RunBuildException buildException = new RunBuildException(e);
            buildException.setLogStacktrace(false);
            throw buildException;
        }

        return createProgramCommandline(toolPath, arguments);
    }

    @NotNull
    @Override
    public List<ProcessListener> getListeners() {
        final CargoLoggerFactory loggerFactory = new CargoLoggerFactory(getLogger());
        return Collections.<ProcessListener>singletonList(new CargoLoggingListener(loggerFactory));
    }
}

File: plugin-rust-agent/src/main/kotlin/jetbrains/buildServer/rust/CargoRunnerBuildService.kt
package jetbrains.buildServer.rust

import com.github.zafarkhaja.semver.Version
import jetbrains.buildServer.RunBuildException
import jetbrains.buildServer.agent.BuildFinishedStatus
import jetbrains.buildServer.agent.BuildRunnerContextEx
import jetbrains.buildServer.agent.inspections.InspectionReporter
import jetbrains.buildServer.agent.runner.ProcessListener
import jetbrains.buildServer.agent.runner.ProgramCommandLine
import jetbrains.buildServer.rust.cargo.*
import jetbrains.buildServer.rust.inspections.ClippyInspectionsParser
import jetbrains.buildServer.rust.inspections.ClippyListener
import jetbrains.buildServer.rust.logging.CargoLoggerFactory
import jetbrains.buildServer.rust.logging.CargoLoggingListener
import jetbrains.buildServer.util.OSType
import jetbrains.buildServer.util.StringUtil
import java.io.File

/**
 * Cargo runner service.
 */
class CargoRunnerBuildService(
        private val runnerContext: BuildRunnerContextEx,
        private val inspectionReporter: InspectionReporter,
        private val clippyInspectionsParser: ClippyInspectionsParser
) : BaseCargoBuildService() {

    private val myCargoWithStdErrVersion = Version.forIntegers(0, 13)
    private val myArgumentsProviders = mapOf(
            CargoConstants.COMMAND_BENCH to BenchArgumentsProvider(),
            CargoConstants.COMMAND_BUILD to BuildArgumentsProvider(),
            CargoConstants.COMMAND_CHECK to CheckArgumentsProvider(),
            CargoConstants.COMMAND_CLEAN to CleanArgumentsProvider(),
            CargoConstants.COMMAND_CLIPPY to ClippyArgumentsProvider(),
            CargoConstants.COMMAND_DOC to DocArgumentsProvider(),
            CargoConstants.COMMAND_LOGIN to LoginArgumentsProvider(),
            CargoConstants.COMMAND_PACKAGE to PackageArgumentsProvider(),
            CargoConstants.COMMAND_PUBLISH to PublishArgumentsProvider(),
            CargoConstants.COMMAND_RUN to RunArgumentsProvider(),
            CargoConstants.COMMAND_RUSTC to RustcArgumentsProvider(),
            CargoConstants.COMMAND_RUSTDOC to RustDocArgumentsProvider(),
            CargoConstants.COMMAND_TEST to TestArgumentsProvider(),
            CargoConstants.COMMAND_UPDATE to UpdateArgumentsProvider(),
            CargoConstants.COMMAND_YANK to YankArgumentsProvider(),
            CargoConstants.COMMAND_CUSTOM_CRATE to CustomCrateArgumentsProvider()
    )

    override fun getRunResult(exitCode: Int): BuildFinishedStatus {
        if (exitCode == 0) {
            return BuildFinishedStatus.FINISHED_SUCCESS
        }

        val commandName = runnerParameters[CargoConstants.PARAM_COMMAND]
        val argumentsProvider = myArgumentsProviders[commandName]
        if (argumentsProvider == null) {
            val buildException = RunBuildException("Unable to construct arguments for cargo command $commandName")
            buildException.isLogStacktrace = false
            throw buildException
        }

        return if (argumentsProvider.shouldFailBuildIfCommandFailed()) {
            BuildFinishedStatus.FINISHED_FAILED
        } else {
            BuildFinishedStatus.FINISHED_SUCCESS
        }
    }

    override fun makeProgramCommandLine(): ProgramCommandLine {
        val parameters = runnerParameters

        val commandName = parameters[CargoConstants.PARAM_COMMAND]
        if (StringUtil.isEmpty(commandName)) {
            val buildException = RunBuildException("Cargo command name is empty")
            buildException.isLogStacktrace = false
            throw buildException
        }

        val argumentsProvider = myArgumentsProviders[commandName]
        if (argumentsProvider == null) {
            val buildException = RunBuildException("Unable to construct arguments for cargo command $commandName")
            buildException.isLogStacktrace = false
            throw buildException
        }

        val arguments = mutableListOf<String>()
        val (toolPath, toolArgs) = getToolPath()
        arguments += toolArgs
        arguments += argumentsProvider.getArguments(runnerContext)

        parameters[CargoConstants.PARAM_ADDITIONAL_ARGUMENTS]?.let {
            arguments += StringUtil.splitHonorQuotes(it)
        }

        runnerContext.configParameters[CargoConstants.CARGO_CONFIG_NAME]?.let {
            if (Version.valueOf(it) >= myCargoWithStdErrVersion) {
                return when (runnerContext.virtualContext.targetOSType) {
                    OSType.WINDOWS -> {
                        createProgramCommandline("cmd.exe", listOf("/c", toolPath) + arguments + listOf("2>&1"))
                    }
                    else -> {
                        val scriptlet = "$toolPath ${arguments.joinToString(" ", transform = ::escapeSingleQuotes)} 2>&1"
                        val toolName = File(toolPath).name
                        createProgramCommandline("/bin/sh", listOf("-c", scriptlet, toolName))
                    }
                }
            }
        }

        return createProgramCommandline(toolPath, arguments)
    }

    private fun escapeSingleQuotes(arg: String): String {
        val escapedArg = StringBuilder(arg.length)
        for (i in 0 until arg.length) {
            val c = arg[i]
            if (c == '\'' && (i == 0 || arg[i - 1] != '\\')) {
                escapedArg.append('\\')
            }
            escapedArg.append(c)
        }
        return escapedArg.toString()
    }

    override fun isCommandLineLoggingEnabled() = false

    override fun getListeners(): List<ProcessListener> =
        super.getListeners() +
        CargoLoggingListener(CargoLoggerFactory(this.logger)) +
        (if (runnerParameters[CargoConstants.PARAM_COMMAND] != CargoConstants.COMMAND_CLIPPY) emptyList()
         else listOf(ClippyListener(inspectionReporter, clippyInspectionsParser)))

    override val blockName: String
        get() = when (val command = runnerParameters[CargoConstants.PARAM_COMMAND]) {
            CargoConstants.COMMAND_CUSTOM_CRATE -> "cargo ${runnerParameters[CargoConstants.PARAM_CUSTOM_CRATE_COMMAND_NAME]}"
            else -> "cargo $command"
        }
}
